File: lib/external/cpp/Mach1Decode.cpp
//  Mach1 SDK
//  Copyright © 2017 Mach1. All rights reserved.

#include "Mach1Decode.h"
#include "Mach1DecodeCAPI.h"

Mach1Decode::Mach1Decode()
{
	M1obj = Mach1DecodeCAPI_create();
}

Mach1Decode::~Mach1Decode()
{
	Mach1DecodeCAPI_delete(M1obj);
} 

std::vector<float> Mach1Decode::horizonAlgo(float Yaw, float Pitch, float Roll, int bufferSize, int sampleIndex)
{
	float* data = Mach1DecodeCAPI_horizonAlgo(M1obj, Yaw, Pitch, Roll, bufferSize, sampleIndex);
	return std::vector<float>(data, data + 10);
}

std::vector<float> Mach1Decode::horizonPairsAlgo(float Yaw, float Pitch, float Roll, int bufferSize, int sampleIndex)
{
	float* data = Mach1DecodeCAPI_horizonPairsAlgo(M1obj, Yaw, Pitch, Roll, bufferSize, sampleIndex);
	return std::vector<float>(data, data + 6);
}

std::vector<float> Mach1Decode::spatialAlgo(float Yaw, float Pitch, float Roll, int bufferSize, int sampleIndex)
{
	float* data = Mach1DecodeCAPI_spatialAlgo(M1obj, Yaw, Pitch, Roll, bufferSize, sampleIndex);
	return std::vector<float>(data, data + 18);
}

std::vector<float> Mach1Decode::spatialAltAlgo(float Yaw, float Pitch, float Roll, int bufferSize, int sampleIndex)
{
	float* data = Mach1DecodeCAPI_spatialAltAlgo(M1obj, Yaw, Pitch, Roll, bufferSize, sampleIndex);
	return std::vector<float>(data, data + 18);
}

std::vector<float> Mach1Decode::spatialPairsAlgo(float Yaw, float Pitch, float Roll, int bufferSize, int sampleIndex)
{
	float* data = Mach1DecodeCAPI_spatialPairsAlgo(M1obj, Yaw, Pitch, Roll, bufferSize, sampleIndex);
	return std::vector<float>(data, data + 10);
}

void Mach1Decode::setAngularSettingsType(AngularSettingsType type)
{
	Mach1DecodeCAPI_setAngularSettingsType(M1obj, (int)type);
}

void Mach1Decode::beginBuffer()
{
	Mach1DecodeCAPI_beginBuffer(M1obj);
}

void Mach1Decode::endBuffer()
{
	Mach1DecodeCAPI_endBuffer(M1obj);
}

long Mach1Decode::getCurrentTime()
{
	return Mach1DecodeCAPI_getCurrentTime(M1obj);
}

char* Mach1Decode::getLog()
{
	return Mach1DecodeCAPI_getLog(M1obj);
}

File: lib/external/swift/Mach1Decode.swift
import Foundation

enum AngularSettingsType : Int {
    case m1Default = 0
    case m1Unity = 1
    case m1UE = 2
    case m1oFEasyCam = 3
    case m1Android = 4
    case m1iOSPortrait = 5
    case m1iOSLandscape = 6
}

enum Mach1AlgorithmType : Int {
    case m1Spatial = 0
    case m1AltSpatial = 1
    case m1Horizon = 2
    case m1HorizonPairs = 3
    case m1SpatialPairs = 4
}

class Mach1Decode {
    var M1obj : UnsafeMutableRawPointer
    
    init() {
        M1obj = Mach1DecodeCAPI_create()
    }
    
    deinit {
        Mach1DecodeCAPI_delete(M1obj)
    }
    
    func setAngularSettingsType(type: AngularSettingsType) {
        Mach1DecodeCAPI_setAngularSettingsType(M1obj, CInt(type.rawValue))
    }
    
    func setAlgorithmType(newAlgorithmType: Mach1AlgorithmType) {
        Mach1DecodeCAPI_setAlgorithmType(M1obj, CInt(newAlgorithmType.rawValue))
    }

    func setFilterSpeed(filterSpeed: Float) {
        Mach1DecodeCAPI_setFilterSpeed(M1obj, filterSpeed)
    }
    
    func beginBuffer() {
        Mach1DecodeCAPI_beginBuffer(M1obj)
    }
    
    func endBuffer() {
        Mach1DecodeCAPI_endBuffer(M1obj)
    }
    
    func getCurrentTime() -> Int {
        let t = Mach1DecodeCAPI_getCurrentTime(M1obj)
        return t
    }
    
    func getLog() -> String {
        let str =  String(cString: UnsafePointer(Mach1DecodeCAPI_getLog(M1obj)))
        return str
    }
    
    func decode(Yaw: Float, Pitch: Float, Roll: Float, bufferSize: Int = 0, sampleIndex: Int = 0) -> [Float] {
        var array: [Float] = []
        let data = Mach1DecodeCAPI_decode(M1obj, Yaw, Pitch, Roll, &array, CInt(bufferSize), CInt(sampleIndex))
        for i in 0..<18 {
            array.append(Float(data[i]))
        }
        return array
    }
   
}

